File: backend/checkout-api/app/src/main/java/develcode/checkout/domain/shared/entities/Entity.java
package develcode.checkout.domain.shared.entities;

import java.time.Instant;

public abstract class Entity {

    private String id;
    private Instant createdAt;
    private Instant updatedAt;

    protected Entity(String id, Instant createdAt, Instant updatedAt) {
        this.id = id;
        this.createdAt = createdAt;
        this.updatedAt = updatedAt;
    }

    public String getId() {
        return id;
    }

    public Instant getCreatedAt() {
        return createdAt;
    }

    public Instant getUpdatedAt() {
        return updatedAt;
    }

    protected void hashChanged() {
        this.updatedAt = Instant.now();
    }

    protected abstract void validate();
}

File: backend/payment-api/src/domain/shared/entity.ts
export abstract class Entity {
  protected constructor(
    protected readonly id: string,
    private createdAt: Date,
    private updatedAt: Date,
  ) {}

  public getCreatedAt(): Date {
    return this.createdAt;
  }

  public getId(): string {
    return this.id;
  }

  public getUpdatedAt(): Date {
    return this.updatedAt;
  }

  protected hasChanged(): void {
    this.updatedAt = new Date();
  }

  protected abstract validate(): void;
}
